### ISSUE TITLE
[cache] Race between CacheConfig addition and Proxy creation when new member joins cluster (was: Please reopen Issue #9735)

### ISSUE THREAD

**Issue by lprimak** [hazelcast/hazelcast#10727]
I am getting #9735 with 3.9-SNAPSHOT as of June 8th

com.hazelcast.cache.CacheNotExistsException: Couldn't find cache config with name XXX

I will furnish example shortly.
There is definitely a race condition with creating of Cache Proxy and Cache Config retrieval as pointed out by @jerrinot

**Comment by tombujok** [comment]
@lprimak reproducer appreciated

**Comment by lprimak** [comment]
I understand. Timeframe is always an issue though

**Comment by lprimak** [comment]
[jcache-race.zip](https://github.com/hazelcast/hazelcast/files/1103853/jcache-race.zip)

**Comment by lprimak** [comment]
You can run the reproducer with:
```
mvn package exec:java
```
It starts up 2 instances which have a CacheLoaderFactoryImpl instance
This class has serialization sleep statements that triggers the race condition.
If you take out the sleep statement from CacheLoaderFactoryImpl, the error doesn't happen.

It doesn't happen every time, but it happens 80% of the time,
so if you don't see the exception, run it a few more times.
```
Jun 26, 2017 7:44:57 PM com.hazelcast.client.spi.impl.ClientMembershipListener
INFO: hz.client_0 [dev] [3.9-SNAPSHOT] 

Members [2] {
	Member [10.0.1.7]:5701 - b88a77f1-cef0-4a2d-a36e-584a2d4989d0
	Member [10.0.1.7]:5702 - 2644e7e6-239d-425f-8001-e526fbb13a10
}

Jun 26, 2017 7:44:57 PM com.flowlogix.jcache.race.CacheLoaderFactoryImpl writeObject
INFO: Writing CacheLoaderFactoryImpl ...
Jun 26, 2017 7:44:57 PM com.flowlogix.jcache.race.CacheLoaderFactoryImpl readObject
INFO: Reading CacheLoaderFactoryImpl ...
Jun 26, 2017 7:44:57 PM com.hazelcast.system
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] Cluster version set to 3.9
Jun 26, 2017 7:44:57 PM com.hazelcast.internal.cluster.ClusterService
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [10.0.1.7]:5701 - b88a77f1-cef0-4a2d-a36e-584a2d4989d0
	Member [10.0.1.7]:5702 - 2644e7e6-239d-425f-8001-e526fbb13a10 this
]

Jun 26, 2017 7:44:57 PM com.hazelcast.cache.impl.CacheService
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] Added cache config: CacheConfig{name='myCache', managerPrefix='/hz/', inMemoryFormat=BINARY, backupCount=1, hotRestart=HotRestartConfig{enabled=false, fsync=false}}
[WARNING] 
com.hazelcast.cache.CacheNotExistsException: Couldn't find cache config with name myCache
	at com.hazelcast.cache.impl.AbstractCacheService.checkCacheSimpleConfig(AbstractCacheService.java:217)
	at com.hazelcast.cache.impl.AbstractCacheService.createDistributedObject(AbstractCacheService.java:197)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.doCreateProxy(ProxyRegistry.java:191)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.createProxy(ProxyRegistry.java:184)
	at com.hazelcast.spi.impl.proxyservice.impl.operations.PostJoinProxyOperation$1.run(PostJoinProxyOperation.java:63)
	at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Jun 26, 2017 7:44:57 PM com.hazelcast.internal.partition.impl.MigrationManager
INFO: [10.0.1.7]:5701 [dev] [3.9-SNAPSHOT] Re-partitioning cluster data... Migration queue size: 271
Jun 26, 2017 7:44:58 PM com.hazelcast.nio.tcp.TcpIpAcceptor
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] Accepting socket connection from /10.0.1.7:56530
Jun 26, 2017 7:44:58 PM com.hazelcast.nio.tcp.TcpIpConnectionManager
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] Established socket connection between /10.0.1.7:5702 and /10.0.1.7:56530
Jun 26, 2017 7:44:58 PM com.hazelcast.client.impl.protocol.task.AuthenticationMessageTask
INFO: [10.0.1.7]:5702 [dev] [3.9-SNAPSHOT] Received auth from Connection[id=2, /10.0.1.7:5702->/10.0.1.7:56530, endpoint=null, alive=true, type=JAVA_CLIENT], successfully authenticated, principal: ClientPrincipal{uuid='74e4bbcf-d7a1-4d04-ac1e-4907a3a78c3f', ownerUuid='b88a77f1-cef0-4a2d-a36e-584a2d4989d0'}, owner connection: false, client version: 3.9-SNAPSHOT
Jun 26, 2017 7:44:58 PM com.hazelcast.client.connection.ClientConnectionManager
```

Thanks!

**Comment by lprimak** [comment]
I did an (incorrect) workaround by doing a number of retries but obviously race condition needs to be fixed at the root cause, as this workaround is clearly temporary

Here is the diff:

```
diff --git a/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java b/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java
index 847cfbfb80..51e055a172 100644
--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.spi.impl.proxyservice.impl;
 
+import com.hazelcast.cache.CacheNotExistsException;
 import com.hazelcast.core.DistributedObject;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.spi.AbstractDistributedObject;
@@ -186,9 +187,18 @@ public final class ProxyRegistry {
 
     private DistributedObjectFuture doCreateProxy(String name, boolean publishEvent, boolean initialize,
                                                   DistributedObjectFuture proxyFuture) {
-        DistributedObject proxy;
+        DistributedObject proxy = null;
         try {
-            proxy = service.createDistributedObject(name);
+            for(int numTries = 1; proxy == null; ++numTries) {
+                try {
+                    proxy = service.createDistributedObject(name);
+                } catch (CacheNotExistsException cnee) {
+                    if(numTries == 10) {
+                        throw cnee;
+                    }
+                    Thread.sleep(50);
+                }
+            }
             if (initialize && proxy instanceof InitializingObject) {
                 try {
                     ((InitializingObject) proxy).initialize();
```

**Comment by lprimak** [comment]
@tombujok this is a pretty serious issue for Payara because JCache isn't working because of this
thanks!

**Comment by vbekiaris** [comment]
The race exists due to `CoreService`s post-join ops being executed before those of other services. Thus `PostJoinProxyOperation` (generated by the `ProxyService` which is a `CoreService`) will be executed before `PostJoinCacheOperation`. Proxy creation within `PostJoinProxyOperation` is asynchronous (the operation just submits the proxy creation `Runnable`s to the system executor), so the race may or may not be observed.
This way, it may be the case that `CacheService` will be requested to create proxies for which `CacheConfig`s are not yet known (as they will become known after `PostJoinCacheOperation` has been executed) --> throws `CacheNotExistsException`.

**Comment by vbekiaris** [comment]
Hi @lprimak , this issue is now fixed in the latest `3.9-SNAPSHOT` by #10902 . Thanks once more for the reproducer.
